File: control/src/tictactoe_computing.cpp
#include <opencv2/opencv.hpp>
#include <iostream>
#include <vector>
#include <algorithm>
#include <random>

class Tic {
private:
    std::vector<int> squares;
    const std::vector<std::vector<int>> winning_combos = {
        {0, 1, 2}, {3, 4, 5}, {6, 7, 8},
        {0, 3, 6}, {1, 4, 7}, {2, 5, 8},
        {0, 4, 8}, {2, 4, 6}
    };
public:
    Tic() : squares(9, -1) {} // Initialize squares to -1
    
    void show() {
        for (size_t i = 0; i < 9; i += 3) {
            for (size_t j = 0; j < 3; ++j) {
                std::cout << squares[i + j] << " ";
            }
            std::cout << std::endl;
        }
    }

    std::vector<int> available_moves() {
        std::vector<int> moves;
        for (size_t i = 0; i < squares.size(); ++i) {
            if (squares[i] == -1) {
                moves.push_back(i);
            }
        }
        return moves;
    }

    std::vector<int> get_squares(int player) {
        std::vector<int> player_squares;
        for (size_t i = 0; i < squares.size(); ++i) {
            if (squares[i] == player) {
                player_squares.push_back(i);
            }
        }
        return player_squares;
    }

    bool complete() {
        if (std::find(squares.begin(), squares.end(), -1) == squares.end()) {
            return true; // All squares filled
        }
        return winner() != -1;
    }

    int winner() {
        for (int player = 0; player <= 1; ++player) {
            std::vector<int> positions = get_squares(player);
            for (const auto& combo : winning_combos) {
                bool win = true;
                for (int pos : combo) {
                    if (std::find(positions.begin(), positions.end(), pos) == positions.end()) {
                        win = false;
                        break;
                    }
                }
                if (win) return player;
            }
        }
        return -1; // No winner yet
    }

    void make_move(int position, int player) {
        squares[position] = player;
    }

    int alphabeta(Tic& node, int player, int alpha, int beta) {
        if (node.complete()) {
            if (node.winner() == 1) return -1;
            else if (node.winner() == 0) return 0;
            else if (node.winner() == 2) return 1;
        }

        for (int move : node.available_moves()) {
            node.make_move(move, player);
            int val = alphabeta(node, get_enemy(player), alpha, beta);
            node.make_move(move, -1);
            if (player == 1) {
                alpha = std::max(alpha, val);
                if (alpha >= beta) return beta;
            } else {
                beta = std::min(beta, val);
                if (beta <= alpha) return alpha;
            }
        }
        return (player == 1) ? alpha : beta;
    }

    int determine(int player) {
        int a = -2;
        std::vector<int> choices;
        if (available_moves().size() == 9) return 4;
        for (int move : available_moves()) {
            make_move(move, player);
            int val = alphabeta(*this, get_enemy(player), -2, 2);
            make_move(move, -1);
            if (val > a) {
                a = val;
                choices = {move};
            } else if (val == a) {
                choices.push_back(move);
            }
        }
        std::random_device rd;
        std::mt19937 gen(rd());
        std::uniform_int_distribution<> dis(0, choices.size() - 1);
        return choices[dis(gen)];
    }

    int get_enemy(int player) {
        return (player == 0) ? 1 : 0;
    }
};

void draw_3x3_grid(cv::Mat& image, std::vector<cv::Point>& contour) {
    cv::Rect boundingRect = cv::boundingRect(contour);

    int cell_width = boundingRect.width / 3;
    int cell_height = boundingRect.height / 3;

    for (int i = 1; i < 3; ++i) {
        cv::line(image, cv::Point(boundingRect.x, boundingRect.y + i * cell_height),
                 cv::Point(boundingRect.x + boundingRect.width, boundingRect.y + i * cell_height),
                 cv::Scalar(0, 255, 0), 2);
    }

    for (int i = 1; i < 3; ++i) {
        cv::line(image, cv::Point(boundingRect.x + i * cell_width, boundingRect.y),
                 cv::Point(boundingRect.x + i * cell_width, boundingRect.y + boundingRect.height),
                 cv::Scalar(0, 255, 0), 2);
    }

    int cell_number = 1;
    for (int i = 0; i < 3; ++i) {
        for (int j = 0; j < 3; ++j) {
            int text_x = boundingRect.x + j * cell_width + cell_width / 2 - 10;
            int text_y = boundingRect.y + i * cell_height + cell_height / 2 + 10;
            cv::putText(image, std::to_string(cell_number), cv::Point(text_x, text_y), cv::FONT_HERSHEY_SIMPLEX, 1, cv::Scalar(0, 0, 255), 2);
            cell_number++;
        }
    }
}

int main() {
    cv::VideoCapture cap(0);
    cap.set(cv::CAP_PROP_FRAME_WIDTH, 640);
    cap.set(cv::CAP_PROP_FRAME_HEIGHT, 480);

    Tic board;
    bool player_turn = true;

    while (true) {
        cv::Mat frame;
        cap.read(frame);

        cv::resize(frame, frame, cv::Size(), 0.5, 0.5);
        cv::Mat gray, blur, thresh;
        cv::cvtColor(frame, gray, cv::COLOR_BGR2GRAY);
        cv::GaussianBlur(gray, blur, cv::Size(5, 5), 0);
        cv::adaptiveThreshold(blur, thresh, 255, 1, 1, 19, 2);

        std::vector<std::vector<cv::Point>> contours;
        cv::findContours(thresh, contours, cv::RETR_EXTERNAL, cv::CHAIN_APPROX_SIMPLE);

        std::vector<std::vector<cv::Point>> filtered_contours;
        for (const auto& contour : contours) {
            double area = cv::contourArea(contour);
            if (area > 1900) {
                std::vector<cv::Point> approx;
                cv::approxPolyDP(contour, approx, 0.1 * cv::arcLength(contour, true), true);
                if (approx.size() == 4) {
                    filtered_contours.push_back(contour);
                    draw_3x3_grid(frame, contour);
                }
            }
        }

        if (player_turn) {
            cv::Mat hsv;
            cv::cvtColor(frame, hsv, cv::COLOR_BGR2HSV);
            cv::Mat mask;
            cv::inRange(hsv, cv::Scalar(0, 100, 100), cv::Scalar(10, 255, 255), mask);

            std::vector<std::vector<cv::Point>> contours;
            cv::findContours(mask, contours, cv::RETR_EXTERNAL, cv::CHAIN_APPROX_SIMPLE);

            if (!contours.empty()) {
                std::vector<cv::Point> largest_contour = *std::max_element(contours.begin(), contours.end(),
                    [](const std::vector<cv::Point>& a, const std::vector<cv::Point>& b) {
                        return cv::contourArea(a) < cv::contourArea(b);
                    });

                cv::Moments M = cv::moments(largest_contour);
                if (M.m00 != 0) {
                    int cx = static_cast<int>(M.m10 / M.m00);
                    int cy = static_cast<int>(M.m01 / M.m00);
                    int cell_row = cy / (frame.rows / 3);
                    int cell_col = cx / (frame.cols / 3);
                    int move = cell_row * 3 + cell_col;
                    if (move >= 0 && move < 9 && board.available_moves()[move]) {
                        board.make_move(move, 0);
                        player_turn = false;
                    }
                }
            }
        } else {
            int move = board.determine(1);
            if (board.available_moves()[move]) {
                board.make_move(move, 1);
                player_turn = true;
            }
        }

        for (size_t i = 0; i < 3; ++i) {
            for (size_t j = 0; j < 3; ++j) {
                if (board.available_moves()[i * 3 + j] == 0) {
                    int cell_x = j * (frame.cols / 3) + (frame.cols / 6);
                    int cell_y = i * (frame.rows / 3) + (frame.rows / 6);
                    cv::circle(frame, cv::Point(cell_x, cell_y), 30, cv::Scalar(0, 0, 255), 3);
                } else if (board.available_moves()[i * 3 + j] == 1) {
                    int cell_x = j * (frame.cols / 3) + (frame.cols / 6);
                    int cell_y = i * (frame.rows / 3) + (frame.rows / 6);
                    cv::circle(frame, cv::Point(cell_x, cell_y), 30, cv::Scalar(255, 0, 0), 3);
                }
            }
        }

        if (board.complete()) {
            int winner = board.winner();
            if (winner == 0) {
                cv::putText(frame, "Player wins!", cv::Point(50, 50), cv::FONT_HERSHEY_SIMPLEX, 1, cv::Scalar(0, 0, 255), 2);
            } else if (winner == 1) {
                cv::putText(frame, "AI wins!", cv::Point(50, 50), cv::FONT_HERSHEY_SIMPLEX, 1, cv::Scalar(255, 0, 0), 2);
            } else {
                cv::putText(frame, "It's a draw!", cv::Point(50, 50), cv::FONT_HERSHEY_SIMPLEX, 1, cv::Scalar(0, 255, 255), 2);
            }
            break;
        }

        cv::imshow("Tic Tac Toe", frame);
        if (cv::waitKey(1) == 'q') break;
    }

    cap.release();
    cv::destroyAllWindows();

    return 0;
}

File: control/src/tictactoe_computing.py
import cv2
import numpy as np
import random

# Function to draw a 3x3 grid on the image and number the cells
def draw_3x3_grid(image, contour):
    # Get the bounding box of the contour
    x, y, w, h = cv2.boundingRect(contour)
    
    # Calculate the width and height of each cell
    cell_width = w // 3
    cell_height = h // 3
    
    # Draw horizontal lines
    for i in range(1, 3):
        cv2.line(image, (x, y + i * cell_height), (x + w, y + i * cell_height), (0, 255, 0), 2)
    
    # Draw vertical lines
    for i in range(1, 3):
        cv2.line(image, (x + i * cell_width, y), (x + i * cell_width, y + h), (0, 255, 0), 2)
    
    # Number the cells from top-left corner
    cell_number = 1
    for i in range(3):
        for j in range(3):
            # Calculate the position for text label
            text_x = x + j * cell_width + cell_width // 2 - 10
            text_y = y + i * cell_height + cell_height // 2 + 10
            # Add cell number
            # Increment cell number
            cell_number += 1

# Function to check if a player has won
def check_winner(board, player):
    for i in range(3):
        if board[i][0] == board[i][1] == board[i][2] == player:
            return True
        if board[0][i] == board[1][i] == board[2][i] == player:
            return True
    if board[0][0] == board[1][1] == board[2][2] == player:
        return True
    if board[0][2] == board[1][1] == board[2][0] == player:
        return True
    return False

class Tic(object):

    winning_combos = (
        [0, 1, 2], [3, 4, 5], [6, 7, 8],
        [0, 3, 6], [1, 4, 7], [2, 5, 8],
        [0, 4, 8], [2, 4, 6])
    winners = ('X-win', 'Draw', 'O-win')

    def __init__(self, squares=[]):
        """Initialize either custom or default board"""
        if len(squares) == 0:
            self.squares = [None for i in range(9)]
        else:
            self.squares = squares

    def show(self):
        """Print game progress"""
        for element in [
                self.squares[i: i + 3] for i in range(0, len(self.squares), 3)]:
            print(element)

    def available_moves(self):
        return [k for k, v in enumerate(self.squares) if v is None]

    def available_combos(self, player):
        return self.available_moves() + self.get_squares(player)

    def complete(self):
        """Check if game has ended"""
        if None not in [v for v in self.squares]:
            return True
        if self.winner() is not None:
            return True
        return False

    def X_won(self):
        return self.winner() == 'X'

    def O_won(self):
        return self.winner() == 'O'

    def tied(self):
        return self.complete() and self.winner() is None

    def winner(self):
        for player in ('X', 'O'):
            positions = self.get_squares(player)
            for combo in self.winning_combos:
                win = True
                for pos in combo:
                    if pos not in positions:
                        win = False
                if win:
                    return player
        return None

    def get_squares(self, player):
        """Returns squares belonging to a player"""
        return [k for k, v in enumerate(self.squares) if v == player]

    def make_move(self, position, player):
        self.squares[position] = player

    def alphabeta(self, node, player, alpha, beta):
        """Alphabeta algorithm"""
        if node.complete():
            if node.X_won():
                return -1
            elif node.tied():
                return 0
            elif node.O_won():
                return 1

        for move in node.available_moves():
            node.make_move(move, player)
            val = self.alphabeta(node, get_enemy(player), alpha, beta)
            node.make_move(move, None)
            if player == 'O':
                if val > alpha:
                    alpha = val
                if alpha >= beta:
                    return beta
            else:
                if val < beta:
                    beta = val
                if beta <= alpha:
                    return alpha
        return alpha if player == 'O' else beta


def get_enemy(player):
    if player == 'X':
        return 'O'
    return 'X'


def determine(board, player):
    """Determine best possible move"""
    a = -2
    choices = []
    if len(board.available_moves()) == 9:
        return 4
    for move in board.available_moves():
        board.make_move(move, player)
        val = board.alphabeta(board, get_enemy(player), -2, 2)
        board.make_move(move, None)
        if val > a:
            a = val
            choices = [move]
        elif val == a:
            choices.append(move)
    return random.choice(choices)


# Initialize webcam
cap = cv2.VideoCapture(0)

# Set the width and height of the camera frame
cap.set(cv2.CAP_PROP_FRAME_WIDTH, 640)
cap.set(cv2.CAP_PROP_FRAME_HEIGHT, 480)

# Create an empty board
board = Tic()

# Variable to keep track of player's turn
player_turn = True

while True:
    # Read a frame from the webcam
    ret, frame = cap.read()

    if not ret:
        break

    # Resize the frame to be 50% smaller
    frame = cv2.resize(frame, (0, 0), fx=0.5, fy=0.5)

    # Convert frame to grayscale
    gray = cv2.cvtColor(frame, cv2.COLOR_BGR2GRAY)

    # Blur and threshold the frame
    blur = cv2.GaussianBlur(gray, (5, 5), 0)
    thresh = cv2.adaptiveThreshold(blur, 255, 1, 1, 19, 2)  # Increased threshold for grid detection

    # Find contours for the grid
    contours_grid, _ = cv2.findContours(thresh, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)

    # Filter contours based on shape (approximated rectangular)
    filtered_contours = []
    for contour in contours_grid:
        peri = cv2.arcLength(contour, True)
        approx = cv2.approxPolyDP(contour, 0.1 * peri, True)  # Adjusted approximation parameter for tighter tolerance
        if len(approx) == 4:  # Check if the contour has 4 vertices (rectangle)
            area = cv2.contourArea(contour)
            if area > 1900:  # Adjust the area threshold as needed
                filtered_contours.append(contour)
                # Draw rectangles for the grid on the original frame
                x, y, w, h = cv2.boundingRect(contour)
                cv2.rectangle(frame, (x, y), (x + w, y + h), (0, 255, 0), 3)
                # Draw 3x3 grid and number the cells
                draw_3x3_grid(frame, contour)

    # If it's the player's turn, check for player's move
    if player_turn:
        # Convert frame to HSV color space
        hsv = cv2.cvtColor(frame, cv2.COLOR_BGR2HSV)

        # Define range of red color in HSV
        lower_red1 = np.array([0, 50, 50])
        upper_red1 = np.array([10, 255, 255])
        lower_red2 = np.array([170, 50, 50])
        upper_red2 = np.array([180, 255, 255])

        # Threshold the HSV image to get only red colors
        mask_red1 = cv2.inRange(hsv, lower_red1, upper_red1)
        mask_red2 = cv2.inRange(hsv, lower_red2, upper_red2)
        mask = cv2.bitwise_or(mask_red1, mask_red2)

        # Find contours in the mask
        contours, _ = cv2.findContours(mask, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)

        cv2.drawContours(frame, contours, -1, (0, 255, 0), 3)

        # Iterate through the contours and find the centroid of the largest contour
        if contours:
            # Get the largest contour
            largest_contour = max(contours, key=cv2.contourArea)
            # Calculate centroid using image moments
            M = cv2.moments(largest_contour)
            if M["m00"] != 0:
                cx = int(M["m10"] / M["m00"])
                cy = int(M["m01"] / M["m00"])
                # Iterate through the cells of the 3x3 grid
                for i in range(3):
                    for j in range(3):
                        # Calculate the top-left corner coordinates of the cell
                        cell_x = x + j * (w // 3)
                        cell_y = y + i * (h // 3)
                        # Calculate the bottom-right corner coordinates of the cell
                        cell_x_end = cell_x + (w // 3)
                        cell_y_end = cell_y + (h // 3)
                        # Check if the centroid falls within the current cell
                        if cell_x < cx < cell_x_end and cell_y < cy < cell_y_end:
                            # Check if the cell is empty
                            if board.squares[i * 3 + j] is None:
                                board.squares[i * 3 + j] = 'X'  # Player's move
                                player_turn = False
                                break


    else:
        # AI's move
        move = determine(board, 'O')
        if board.squares[move] is None:
            board.squares[move] = 'O'  # AI's move
            player_turn = True
    
    # Draw player's moves (red circles) and AI's moves (blue circles)
    for i in range(3):
        for j in range(3):
            if board.squares[i * 3 + j] == 'X':
                cell_x = x + j * (w // 3) + (w // 3) // 2
                cell_y = y + i * (h // 3) + (h // 3) // 2
                cv2.circle(frame, (cell_x, cell_y), 30, (0, 0, 255), 3)  # Player's move (red circle)
            elif board.squares[i * 3 + j] == 'O':
                cell_x = x + j * (w // 3) + (w // 3) // 2
                cell_y = y + i * (h // 3) + (h // 3) // 2
                cv2.circle(frame, (cell_x, cell_y), 30, (255, 0, 0), 3)  # AI's move (blue circle)
    
    # Check for a winner
    if board.complete():
        winner = board.winner()
        if winner == 'X':
            cv2.putText(frame, "Player wins!", (50, 50), cv2.FONT_HERSHEY_SIMPLEX, 1, (0, 0, 255), 2)
        elif winner == 'O':
            cv2.putText(frame, "AI wins!", (50, 50), cv2.FONT_HERSHEY_SIMPLEX, 1, (255, 0, 0), 2)
        else:
            cv2.putText(frame, "It's a draw!", (50, 50), cv2.FONT_HERSHEY_SIMPLEX, 1, (0, 255, 255), 2)
        break

    # Display the frame
    cv2.imshow("Tic Tac Toe", frame)

    # Break the loop when 'q' is pressed
    if cv2.waitKey(1) & 0xFF == ord('q'):
        break

# Release the webcam and close all windows
cap.release()
cv2.destroyAllWindows()
